runtime/builtin_types.go

```diff
@@ -248,6 +249,82 @@ func builtinMapFn(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return NewList(result...).ToObject(), nil
 }
 
+func builtinFilter(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc < 2 {
+		return nil, f.RaiseType(TypeErrorType, fmt.Sprintf("filter expected 2 arguments, got %d", argc))
+	}
+	fn := args[0]
+	l := args[1]
+	switch {
+	case l.isInstance(TupleType):
+		result := make([]*Object, 0)
+		for _, item := range toTupleUnsafe(l).elems {
+			if fn == None {
+				if ret, raised := IsTrue(f, item); raised != nil {
+					return nil, raised
+				} else if ret {
+					result = append(result, item)
+				}
+			} else {
+				ret, raised := args[0].Call(f, []*Object{item}, nil)
+				if raised != nil {
+					return nil, raised
+				}
+				if ret, raised := IsTrue(f, ret); raised != nil {
+					return nil, raised
+				} else if ret {
+					result = append(result, item)
+				}
+			}
+		}
+		return NewTuple(result...).ToObject(), nil
+	case l.isInstance(StrType):
+		if fn == None {
+			return l, nil
+		}
+		var result bytes.Buffer
+		for _, item := range toStrUnsafe(l).Value() {
```

Comment:
This will iterate over l assuming it's a UTF-8 encoded string. I think what you actually want is to convert this to a []byte and iterate over each byte and WriteByte() below instead of WriteRune().